### PR TITLE
[SLP]Remove operands upon marking instruction for deletion.

### PR DESCRIPTION
**Pull request by alexey-bataev**
If the instruction is marked for deletion, better to drop all its
operands and mark them for deletion too (if allowed). It allows to have
more vectorizable patterns and generate less useless extractelement
instructions.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 1]
```diff
@@ -17344,14 +17370,11 @@ class HorizontalReduction {
...
-                    R, cast<Instruction>(ReductionOps.front().front())
-                           ->getModule()
-                           ->getDataLayout());
-                return !Known.isNonNegative();
-              }));
+          assert(ReducedSubTree->getType() != VL.front()->getType() &&
+                 "Expected different reduction type.");
+          ReducedSubTree =
+              Builder.CreateIntCast(ReducedSubTree, VL.front()->getType(),
+                                    *V.isSignedMinBitwidthRootNode());
```
Comment (RKSimon) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:17377]: Why is it safe to directly dereference an optional?

Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp]: I reworked the function, if the types do not match, it means that the root node is resized due to minbitwidth analysis and there is signedness info in MinBWs map for the root node

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 2]
```diff
@@ -14064,11 +14070,24 @@ Value *BoUpSLP::vectorizeTree(
...
+      for (unsigned Idx : seq<unsigned>(I->getNumOperands())) {
+        // Ignore pointer operand of stores to keep correct DIAssignID.
+        if (isa<StoreInst>(I) && Idx == 1)
+          continue;
+        Value *Op = I->getOperand(Idx);
+        I->setOperand(Idx, PoisonValue::get(Op->getType()));
+        if (auto *OpI = dyn_cast<Instruction>(Op))
+          if (!isDeleted(OpI) && isInstructionTriviallyDead(OpI, TLI) &&
+              Entry->VectorizedValue != OpI)
+            eraseInstruction(OpI);
```
Comment (RKSimon) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:14084]: Similar code to the version in HorizontalReduction - pull out into a small helper?

**Review by RKSimon — APPROVED** [review]
LGTM with one minor

**Comment by dyung** [comment]
Hi @alexey-bataev, we are seeing an assertion failure in internal tests which I bisected back to this change. What is odd is that it does not hit the assertion failure reliably, so there seems to be some element of randomness to it. I have put the repro and assertion failure stack trace in issue #98133. Can you take a look?

**Comment by mstorsjo** [comment]
This change has caused failed asserts. Strangely enough, the failed asserts aren't entirely deterministic though. Reduced testcase:

```c
int b;
float c;
int e;
void f(float p1[]) {
  float *j = &b;
  for (e = 0; e < 2;) {
    float a;
    float d;
    {
      a += p1[e] * 0;
      d += e * j[e];
    }
    e = a;
    c = d;
  }
}
```

To reproduce:
```
$ clang -target i686-linux-gnu -c repro.c -O2 -g
clang: ../lib/IR/Instruction.cpp:153: void llvm::Instruction::insertBefore(llvm:
:BasicBlock&, llvm::iplist_impl<llvm::simple_ilist<llvm::Instruction, llvm::ilis
t_iterator_bits<true>, llvm::ilist_parent<llvm::BasicBlock> >, llvm::SymbolTable
ListTraits<llvm::Instruction, llvm::ilist_iterator_bits<true>, llvm::ilist_paren
t<llvm::BasicBlock> > >::iterator): Assertion `!isa<PHINode>(this) && "Inserting
 PHI after debug-records!"' failed.
```

Please have a look, and consider reverting if fixing takes a long time.

**Comment by OCHyams** [comment]
More info about the assert in case its helpful:

We typically see this assert when a pass sets an insertion point to the beginning of a block or the first non-phi using an `Instruction *` rather than iterator (which carries a bit that describes whether the insertion point comes before the instruction only or before the instruction and the debug records attached to it). Scanning it I couldn't see anything obvious in this patch that would cause that assertion to fire - it's possible that this patch is just exercising a latent bug in SLP. I also have no clues as to why it's nondeterministic.

DbgRecords and the new iterator bit are documented over [here](https://llvm.org/docs/RemoveDIsDebugInfo.html#api-changes). I'm happy to help with any debug-info questions.

**Comment by alexey-bataev** [comment]
Hi, thanks for the reports, will fix ASAP

**Comment by alanzhao1** [comment]
Hi - Chrome is seeing crashes due to this PR as well (https://crbug.com/351866598).

Unfortunately, https://github.com/llvm/llvm-project/commit/2cba218ca573f5da6ff301df59aa686d2e3e40c5 does *not* fix the crash in our case.

**Comment by alexey-bataev** [comment]
> Hi - Chrome is seeing crashes due to this PR as well (https://crbug.com/351866598).
> 
> Unfortunately, [2cba218](https://github.com/llvm/llvm-project/commit/2cba218ca573f5da6ff301df59aa686d2e3e40c5) does _not_ fix the crash in our case.

Can you provide a reproducer?

**Comment by alanzhao1** [comment]
See [repro.tar.gz](https://github.com/user-attachments/files/16149916/repro.tar.gz)

I'll see if I can get a reduced repro.

**Comment by alexey-bataev** [comment]
> See [repro.tar.gz](https://github.com/user-attachments/files/16149916/repro.tar.gz)
> 
> I'll see if I can get a reduced repro.

Ok, able to reproduce, will try to investigate

**Comment by alanzhao1** [comment]
> > See [repro.tar.gz](https://github.com/user-attachments/files/16149916/repro.tar.gz)
> > I'll see if I can get a reduced repro.
> 
> Ok, able to reproduce, will try to investigate

Here's a (much) reduced repro with CReduce:

```c
int a, b, c;
double d, e, f, g;
void h() {
  do {
    while (a)
      ;
    d = f - (0 * f + e * g);
    g = e * f - 0 * g + g;
    f = d * g;
    a = a - c;
  } while (b);
}
```

Reduced command: `clang -cc1 -triple i386-pc-windows-msvc19.34.0 -target-feature +sse3 -O2 -vectorize-slp -emit-llvm fft-06e938.reduced.c`

**Comment by alexey-bataev** [comment]
I reduced already, the fix will be ready soon

**Comment by alexey-bataev** [comment]
Must be fixed in af21bc19172410f29606cac82a96c33d6710fe38

**Comment by alanzhao1** [comment]
> Must be fixed in [af21bc1](https://github.com/llvm/llvm-project/commit/af21bc19172410f29606cac82a96c33d6710fe38)

Confirmed. Thanks for the extremely quick fix!

**Comment by mikaelholmen** [comment]
Another crash bisected to this patch:
```opt -passes=slp-vectorizer bbi-97294.ll -o /dev/null```

It still crashes on current top of tree 015526bf3fc68 .

[bbi-97294.ll.gz](https://github.com/user-attachments/files/16155883/bbi-97294.ll.gz)

**Comment by alexey-bataev** [comment]
Fixed in 3742c2a83cd644c47db969b6dd421665695f94da

**Comment by mikaelholmen** [comment]
> Fixed in [3742c2a](https://github.com/llvm/llvm-project/commit/3742c2a83cd644c47db969b6dd421665695f94da)

Confirmed that it fixes the crash.

**Comment by alexey-bataev** [comment]
Need a reproducer!

**Comment by alexfh** [comment]
Just got to this, which crashes `opt -passes=slp-vectorizer`:
```
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-generic-linux-gnu"

define void @_h() #0 {
entry:
  %arrayidx51.1.i = getelementptr i8, ptr null, i64 4
  %retval.sroa.3.0.insert.ext.i.i = zext i8 0 to i32
  %retval.sroa.2.0.insert.ext.i.i = zext i8 0 to i32
  br label %cond.end.i

cond.end.i:                                       ; preds = %entry
  %add46.i30 = or i32 0, %retval.sroa.2.0.insert.ext.i.i
  %add49.i = or i32 0, %retval.sroa.3.0.insert.ext.i.i
    #dbg_value(i32 %add49.i, !8, !DIExpression(), !16)
  %0 = tail call i32 @llvm.umin.i32(i32 %add46.i30, i32 0)
  %cmp.i14.i.i.i = icmp slt i32 %add49.i, 0
  %block_color.sroa.7.0.insert.ext.i = select i1 %cmp.i14.i.i.i, i32 0, i32 0
  %block_color.sroa.7.0.insert.shift.i = shl i32 %block_color.sroa.7.0.insert.ext.i, 16
  %block_color.sroa.5.0.insert.ext.i = select i1 false, i32 0, i32 %0
  %block_color.sroa.5.0.insert.shift.i = shl i32 %block_color.sroa.5.0.insert.ext.i, 0
  %block_color.sroa.7.0.insert.insert.i = or i32 %block_color.sroa.7.0.insert.shift.i, %block_color.sroa.5.0.insert.shift.i
  %block_color.sroa.5.0.insert.insert.i = or i32 %block_color.sroa.7.0.insert.insert.i, 0
  %block_color.sroa.0.0.insert.insert.i = or i32 %block_color.sroa.5.0.insert.insert.i, 0
  store i32 %block_color.sroa.0.0.insert.insert.i, ptr null, align 4
  %add46.1.i = or i32 0, %retval.sroa.2.0.insert.ext.i.i
  %add49.1.i = or i32 0, %retval.sroa.3.0.insert.ext.i.i
  %cmp.i11.i.i.1.i = icmp slt i32 %add46.1.i, 0
  %1 = tail call i32 @llvm.umin.i32(i32 %add49.1.i, i32 0)
  %block_color.sroa.7.0.insert.ext.1.i = select i1 false, i32 0, i32 %1
  %block_color.sroa.7.0.insert.shift.1.i = shl i32 %block_color.sroa.7.0.insert.ext.1.i, 16
  %block_color.sroa.5.0.insert.ext.1.i = select i1 %cmp.i11.i.i.1.i, i32 0, i32 0
  %block_color.sroa.5.0.insert.shift.1.i = shl i32 %block_color.sroa.5.0.insert.ext.1.i, 0
  %block_color.sroa.7.0.insert.insert.1.i = or i32 %block_color.sroa.7.0.insert.shift.1.i, %block_color.sroa.5.0.insert.shift.1.i
  %block_color.sroa.5.0.insert.insert.1.i = or i32 %block_color.sroa.7.0.insert.insert.1.i, 0
  %block_color.sroa.0.0.insert.insert.1.i = or i32 %block_color.sroa.5.0.insert.insert.1.i, 0
  store i32 %block_color.sroa.0.0.insert.insert.1.i, ptr %arrayidx51.1.i, align 4
  ret void
}

; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
declare i32 @llvm.umin.i32(i32, i32) #1

; uselistorder directives
uselistorder ptr @llvm.umin.i32, { 1, 0 }

attributes #0 = { "target-features"="+aes,+cmov,+crc32,+cx16,+cx8,+fxsr,+mmx,+pclmul,+popcnt,+prfchw,+sse,+sse2,+sse3,+sse4.1,+sse4.2,+ssse3,+x87" }
attributes #1 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }

!llvm.dbg.cu = !{!0, !3, !5}
!llvm.module.flags = !{!7}

!0 = distinct !DICompileUnit(language: DW_LANG_C_plus_plus_14, file: !1, producer: "clang 9ddfe62f5c11e3f65f444209f514029ded2d58b9", isOptimized: true, runtimeVersion: 0, splitDebugFilename: "q.pic.dwo", emissionKind: FullDebug, enums: !2, retainedTypes: !2, globals: !2, imports: !2, splitDebugInlining: false, debugInfoForProfiling: true, nameTableKind: GNU)
!1 = !DIFile(filename: "q.cpp", directory: "/proc/self/cwd", checksumkind: CSK_MD5, checksum: "7f405f2fa7a85732c62eeb5ad41e9a99")
!2 = !{}
!3 = distinct !DICompileUnit(language: DW_LANG_C_plus_plus_14, file: !4, producer: "clang 9ddfe62f5c11e3f65f444209f514029ded2d58b9", isOptimized: true, runtimeVersion: 0, splitDebugFilename: "q.pic.dwo", emissionKind: FullDebug, splitDebugInlining: false, debugInfoForProfiling: true, nameTableKind: GNU)
!4 = !DIFile(filename: "q.cpp", directory: "/proc/self/cwd", checksumkind: CSK_MD5, checksum: "b1f6e06104b120054aa37f5011b1165d")
!5 = distinct !DICompileUnit(language: DW_LANG_C_plus_plus_14, file: !6, producer: "clang 9ddfe62f5c11e3f65f444209f514029ded2d58b9", isOptimized: true, runtimeVersion: 0, splitDebugFilename: "q.pic.dwo", emissionKind: FullDebug, splitDebugInlining: false, debugInfoForProfiling: true, nameTableKind: GNU)
!6 = !DIFile(filename: "q.cpp", directory: "/proc/self/cwd", checksumkind: CSK_MD5, checksum: "1a3869a68f60205ccd08e442850492ac")
!7 = !{i32 2, !"Debug Info Version", i32 3}
!8 = !DILocalVariable(name: "sb", arg: 4, scope: !9, file: !10, line: 813, type: !15)
!9 = distinct !DISubprogram(name: "color_rgba", linkageName: "_i", scope: !11, file: !10, line: 813, type: !13, scopeLine: 814, flags: DIFlagPrototyped | DIFlagAllCallsDescribed, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !0, declaration: !14, retainedNodes: !2)
!10 = !DIFile(filename: "q.h", directory: "/proc/self/cwd", checksumkind: CSK_MD5, checksum: "798e10c1d58d4180e9266db9b9a5137b")
!11 = distinct !DICompositeType(tag: DW_TAG_class_type, name: "color_rgba", scope: !12, file: !10, line: 763, size: 32, flags: DIFlagTypePassByValue | DIFlagNonTrivial, elements: !2, identifier: "_ZTSN6basisu10color_rgbaE")
!12 = !DINamespace(name: "basisu", scope: null)
!13 = distinct !DISubroutineType(types: !2)
!14 = !DISubprogram(name: "color_rgba", scope: !11, file: !10, line: 813, type: !13, scopeLine: 813, flags: DIFlagPublic | DIFlagPrototyped, spFlags: DISPFlagOptimized)
!15 = !DIBasicType(name: "int", size: 32, encoding: DW_ATE_signed)
!16 = !DILocation(line: 0, scope: !9, inlinedAt: !17)
!17 = distinct !DILocation(line: 1320, column: 16, scope: !18, inlinedAt: !29)
!18 = distinct !DILexicalBlock(scope: !19, file: !1, line: 1318, column: 4)
!19 = distinct !DILexicalBlock(scope: !20, file: !1, line: 1317, column: 4)
!20 = distinct !DILexicalBlock(scope: !21, file: !1, line: 1317, column: 4)
!21 = distinct !DILexicalBlock(scope: !22, file: !1, line: 1312, column: 3)
!22 = distinct !DILexicalBlock(scope: !23, file: !1, line: 1311, column: 3)
!23 = distinct !DILexicalBlock(scope: !24, file: !1, line: 1311, column: 3)
!24 = distinct !DISubprogram(name: "evaluate_solution_fast", linkageName: "_f", scope: !25, file: !1, line: 1276, type: !27, scopeLine: 1277, flags: DIFlagPrototyped | DIFlagAllCallsDescribed, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !0, declaration: !28, retainedNodes: !2)
!25 = distinct !DICompositeType(tag: DW_TAG_class_type, name: "u", scope: !12, file: !26, line: 924, size: 3072, flags: DIFlagFwdDecl | DIFlagNonTrivial, identifier: "_j")
!26 = !DIFile(filename: "q.h", directory: "/proc/self/cwd", checksumkind: CSK_MD5, checksum: "fd8326e7945b2062220a64fd8351349c")
!27 = distinct !DISubroutineType(types: !2)
!28 = !DISubprogram(name: "evaluate_solution_fast", linkageName: "_f", scope: !25, file: !26, line: 1089, type: !27, scopeLine: 1089, flags: DIFlagPrototyped, spFlags: DISPFlagOptimized)
!29 = distinct !DILocation(line: 1096, column: 12, scope: !30, inlinedAt: !33)
!30 = distinct !DILexicalBlock(scope: !31, file: !26, line: 1093, column: 8)
!31 = distinct !DISubprogram(name: "evaluate_solution", linkageName: "_g", scope: !25, file: !26, line: 1091, type: !27, scopeLine: 1092, flags: DIFlagPrototyped | DIFlagAllCallsDescribed, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !0, declaration: !32, retainedNodes: !2)
!32 = !DISubprogram(name: "evaluate_solution", linkageName: "_g", scope: !25, file: !26, line: 1091, type: !27, scopeLine: 1091, flags: DIFlagPrototyped, spFlags: DISPFlagOptimized)
!33 = distinct !DILocation(line: 990, column: 4, scope: !34)
!34 = !DILexicalBlockFile(scope: !35, file: !1, discriminator: 2)
!35 = distinct !DILexicalBlock(scope: !36, file: !1, line: 958, column: 3)
!36 = distinct !DILexicalBlock(scope: !37, file: !1, line: 957, column: 3)
!37 = distinct !DILexicalBlock(scope: !38, file: !1, line: 957, column: 3)
!38 = distinct !DISubprogram(name: "y", linkageName: "_h", scope: !25, file: !1, line: 947, type: !39, scopeLine: 948, flags: DIFlagPrototyped | DIFlagAllCallsDescribed, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !0, declaration: !40, retainedNodes: !2)
!39 = distinct !DISubroutineType(types: !2)
!40 = !DISubprogram(name: "y", linkageName: "_h", scope: !25, file: !26, line: 1101, type: !39, scopeLine: 1101, flags: DIFlagPrototyped, spFlags: DISPFlagOptimized)
```

**Comment by alexey-bataev** [comment]
> Just got to this, which crashes `opt -passes=slp-vectorizer`:
> 
> ```
> target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
> target triple = "x86_64-generic-linux-gnu"
> 
> define void @_h() #0 {
> entry:
>   %arrayidx51.1.i = getelementptr i8, ptr null, i64 4
>   %retval.sroa.3.0.insert.ext.i.i = zext i8 0 to i32
>   %retval.sroa.2.0.insert.ext.i.i = zext i8 0 to i32
>   br label %cond.end.i
> 
> cond.end.i:                                       ; preds = %entry
>   %add46.i30 = or i32 0, %retval.sroa.2.0.insert.ext.i.i
>   %add49.i = or i32 0, %retval.sroa.3.0.insert.ext.i.i
>     #dbg_value(i32 %add49.i, !8, !DIExpression(), !16)
>   %0 = tail call i32 @llvm.umin.i32(i32 %add46.i30, i32 0)
>   %cmp.i14.i.i.i = icmp slt i32 %add49.i, 0
>   %block_color.sroa.7.0.insert.ext.i = select i1 %cmp.i14.i.i.i, i32 0, i32 0
>   %block_color.sroa.7.0.insert.shift.i = shl i32 %block_color.sroa.7.0.insert.ext.i, 16
>   %block_color.sroa.5.0.insert.ext.i = select i1 false, i32 0, i32 %0
>   %block_color.sroa.5.0.insert.shift.i = shl i32 %block_color.sroa.5.0.insert.ext.i, 0
>   %block_color.sroa.7.0.insert.insert.i = or i32 %block_color.sroa.7.0.insert.shift.i, %block_color.sroa.5.0.insert.shift.i
>   %block_color.sroa.5.0.insert.insert.i = or i32 %block_color.sroa.7.0.insert.insert.i, 0
>   %block_color.sroa.0.0.insert.insert.i = or i32 %block_color.sroa.5.0.insert.insert.i, 0
>   store i32 %block_color.sroa.0.0.insert.insert.i, ptr null, align 4
>   %add46.1.i = or i32 0, %retval.sroa.2.0.insert.ext.i.i
>   %add49.1.i = or i32 0, %retval.sroa.3.0.insert.ext.i.i
>   %cmp.i11.i.i.1.i = icmp slt i32 %add46.1.i, 0
>   %1 = tail call i32 @llvm.umin.i32(i32 %add49.1.i, i32 0)
>   %block_color.sroa.7.0.insert.ext.1.i = select i1 false, i32 0, i32 %1
>   %block_color.sroa.7.0.insert.shift.1.i = shl i32 %block_color.sroa.7.0.insert.ext.1.i, 16
>   %block_color.sroa.5.0.insert.ext.1.i = select i1 %cmp.i11.i.i.1.i, i32 0, i32 0
>   %block_color.sroa.5.0.insert.shift.1.i = shl i32 %block_color.sroa.5.0.insert.ext.1.i, 0
>   %block_color.sroa.7.0.insert.insert.1.i = or i32 %block_color.sroa.7.0.insert.shift.1.i, %block_color.sroa.5.0.insert.shift.1.i
>   %block_color.sroa.5.0.insert.insert.1.i = or i32 %block_color.sroa.7.0.insert.insert.1.i, 0
>   %block_color.sroa.0.0.insert.insert.1.i = or i32 %block_color.sroa.5.0.insert.insert.1.i, 0
>   store i32 %block_color.sroa.0.0.insert.insert.1.i, ptr %arrayidx51.1.i, align 4
>   ret void
> }
> 
> ; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
> declare i32 @llvm.umin.i32(i32, i32) #1
> 
> ; uselistorder directives
> uselistorder ptr @llvm.umin.i32, { 1, 0 }
> 
> attributes #0 = { "target-features"="+aes,+cmov,+crc32,+cx16,+cx8,+fxsr,+mmx,+pclmul,+popcnt,+prfchw,+sse,+sse2,+sse3,+sse4.1,+sse4.2,+ssse3,+x87" }
> attributes #1 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }
> 
> !llvm.dbg.cu = !{!0, !3, !5}
> !llvm.module.flags = !{!7}
> 
> !0 = distinct !DICompileUnit(language: DW_LANG_C_plus_plus_14, file: !1, producer: "clang 9ddfe62f5c11e3f65f444209f514029ded2d58b9", isOptimized: true, runtimeVersion: 0, splitDebugFilename: "q.pic.dwo", emissionKind: FullDebug, enums: !2, retainedTypes: !2, globals: !2, imports: !2, splitDebugInlining: false, debugInfoForProfiling: true, nameTableKind: GNU)
> !1 = !DIFile(filename: "q.cpp", directory: "/proc/self/cwd", checksumkind: CSK_MD5, checksum: "7f405f2fa7a85732c62eeb5ad41e9a99")
> !2 = !{}
> !3 = distinct !DICompileUnit(language: DW_LANG_C_plus_plus_14, file: !4, producer: "clang 9ddfe62f5c11e3f65f444209f514029ded2d58b9", isOptimized: true, runtimeVersion: 0, splitDebugFilename: "q.pic.dwo", emissionKind: FullDebug, splitDebugInlining: false, debugInfoForProfiling: true, nameTableKind: GNU)
> !4 = !DIFile(filename: "q.cpp", directory: "/proc/self/cwd", checksumkind: CSK_MD5, checksum: "b1f6e06104b120054aa37f5011b1165d")
> !5 = distinct !DICompileUnit(language: DW_LANG_C_plus_plus_14, file: !6, producer: "clang 9ddfe62f5c11e3f65f444209f514029ded2d58b9", isOptimized: true, runtimeVersion: 0, splitDebugFilename: "q.pic.dwo", emissionKind: FullDebug, splitDebugInlining: false, debugInfoForProfiling: true, nameTableKind: GNU)
> !6 = !DIFile(filename: "q.cpp", directory: "/proc/self/cwd", checksumkind: CSK_MD5, checksum: "1a3869a68f60205ccd08e442850492ac")
> !7 = !{i32 2, !"Debug Info Version", i32 3}
> !8 = !DILocalVariable(name: "sb", arg: 4, scope: !9, file: !10, line: 813, type: !15)
> !9 = distinct !DISubprogram(name: "color_rgba", linkageName: "_i", scope: !11, file: !10, line: 813, type: !13, scopeLine: 814, flags: DIFlagPrototyped | DIFlagAllCallsDescribed, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !0, declaration: !14, retainedNodes: !2)
> !10 = !DIFile(filename: "q.h", directory: "/proc/self/cwd", checksumkind: CSK_MD5, checksum: "798e10c1d58d4180e9266db9b9a5137b")
> !11 = distinct !DICompositeType(tag: DW_TAG_class_type, name: "color_rgba", scope: !12, file: !10, line: 763, size: 32, flags: DIFlagTypePassByValue | DIFlagNonTrivial, elements: !2, identifier: "_ZTSN6basisu10color_rgbaE")
> !12 = !DINamespace(name: "basisu", scope: null)
> !13 = distinct !DISubroutineType(types: !2)
> !14 = !DISubprogram(name: "color_rgba", scope: !11, file: !10, line: 813, type: !13, scopeLine: 813, flags: DIFlagPublic | DIFlagPrototyped, spFlags: DISPFlagOptimized)
> !15 = !DIBasicType(name: "int", size: 32, encoding: DW_ATE_signed)
> !16 = !DILocation(line: 0, scope: !9, inlinedAt: !17)
> !17 = distinct !DILocation(line: 1320, column: 16, scope: !18, inlinedAt: !29)
> !18 = distinct !DILexicalBlock(scope: !19, file: !1, line: 1318, column: 4)
> !19 = distinct !DILexicalBlock(scope: !20, file: !1, line: 1317, column: 4)
> !20 = distinct !DILexicalBlock(scope: !21, file: !1, line: 1317, column: 4)
> !21 = distinct !DILexicalBlock(scope: !22, file: !1, line: 1312, column: 3)
> !22 = distinct !DILexicalBlock(scope: !23, file: !1, line: 1311, column: 3)
> !23 = distinct !DILexicalBlock(scope: !24, file: !1, line: 1311, column: 3)
> !24 = distinct !DISubprogram(name: "evaluate_solution_fast", linkageName: "_f", scope: !25, file: !1, line: 1276, type: !27, scopeLine: 1277, flags: DIFlagPrototyped | DIFlagAllCallsDescribed, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !0, declaration: !28, retainedNodes: !2)
> !25 = distinct !DICompositeType(tag: DW_TAG_class_type, name: "u", scope: !12, file: !26, line: 924, size: 3072, flags: DIFlagFwdDecl | DIFlagNonTrivial, identifier: "_j")
> !26 = !DIFile(filename: "q.h", directory: "/proc/self/cwd", checksumkind: CSK_MD5, checksum: "fd8326e7945b2062220a64fd8351349c")
> !27 = distinct !DISubroutineType(types: !2)
> !28 = !DISubprogram(name: "evaluate_solution_fast", linkageName: "_f", scope: !25, file: !26, line: 1089, type: !27, scopeLine: 1089, flags: DIFlagPrototyped, spFlags: DISPFlagOptimized)
> !29 = distinct !DILocation(line: 1096, column: 12, scope: !30, inlinedAt: !33)
> !30 = distinct !DILexicalBlock(scope: !31, file: !26, line: 1093, column: 8)
> !31 = distinct !DISubprogram(name: "evaluate_solution", linkageName: "_g", scope: !25, file: !26, line: 1091, type: !27, scopeLine: 1092, flags: DIFlagPrototyped | DIFlagAllCallsDescribed, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !0, declaration: !32, retainedNodes: !2)
> !32 = !DISubprogram(name: "evaluate_solution", linkageName: "_g", scope: !25, file: !26, line: 1091, type: !27, scopeLine: 1091, flags: DIFlagPrototyped, spFlags: DISPFlagOptimized)
> !33 = distinct !DILocation(line: 990, column: 4, scope: !34)
> !34 = !DILexicalBlockFile(scope: !35, file: !1, discriminator: 2)
> !35 = distinct !DILexicalBlock(scope: !36, file: !1, line: 958, column: 3)
> !36 = distinct !DILexicalBlock(scope: !37, file: !1, line: 957, column: 3)
> !37 = distinct !DILexicalBlock(scope: !38, file: !1, line: 957, column: 3)
> !38 = distinct !DISubprogram(name: "y", linkageName: "_h", scope: !25, file: !1, line: 947, type: !39, scopeLine: 948, flags: DIFlagPrototyped | DIFlagAllCallsDescribed, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !0, declaration: !40, retainedNodes: !2)
> !39 = distinct !DISubroutineType(types: !2)
> !40 = !DISubprogram(name: "y", linkageName: "_h", scope: !25, file: !26, line: 1101, type: !39, scopeLine: 1101, flags: DIFlagPrototyped, spFlags: DISPFlagOptimized)
> ```

Thanks, will fix ASAP

**Comment by alexey-bataev** [comment]
Fixed in 9e261c5bee4c5618723c243986c4b39236713378

**Comment by alexfh** [comment]
Thanks! I'll let you know if I see any more related issues.

**Pull request merged**